src/storage/src/storage_state.rs

```diff
@@ -261,80 +261,64 @@ impl<'a, A: Allocate, B: StorageCapture> ActiveStorageState<'a, A, B> {
                 }
             }
 
-            StorageCommand::AdvanceAllLocalInputs { advance_to } => {
-                let new_frontier = Antichain::from_elem(advance_to);
-                for (id, table_state) in &mut self.storage_state.table_state {
-                    for local_input in &mut table_state.inputs {
-                        if let Some(capability) = local_input.capability.upgrade() {
-                            capability.borrow_mut().downgrade(&advance_to);
+            StorageCommand::Append(appends) => {
+                for (id, updates, upper) in appends {
```

Comment:
This is just batching things together, otherwise we'll have to be sending as many commands per second as there are tables in order to advance timestamps

Reply:
I agree, but am worried about callers of this command knowing that it's not atomic, and could be partially applied if there's an ill timed restart.

Reply:
Just to check; why wouldn't the `StorageCommand::Append` command be atomic?

Reply:
I thought it was because it's a loop over `appends`, and if a crash happens mid loop, then if some of the table states are backed by persist, then those would be around after the restart but the rest of `appends` would not.

Reply:
Well, worth shaking out! My understanding is that the API provided by `StorageController` methods are meant to be "transactional" in that the method is intended to return without error only once the command is durable (potentially only in the controller state, until it is later acked as received by storage). I think this is currently true for .. zero-ish of the methods, but that this is our goal. So the implementation now certainly isn't atomic, transactional, durable, etc for just about all of the controller commands, but we want to make it so.

Reply:
Yes, the goal is for each command to be transactional and atomic in that when it returns without error you can be certain that it will take effect. When an API offers a batched variant then a single function call packs a lot of subcomands inside it which are subject to the same goals.

The tricky bit is what happens when a batched command function call returns an error. My take is that the caller should assume that none of the supplied commands took effect (but in reality a subset of those might have successfully executed). This is why another property of commands that we wish to have is idempotency so that the caller can simply retry the batch operation until they get a successful response. The subset of commands that had previously gone through would be no-ops and the ones that didn't will execute.

Does this sound like a good idea?

Reply:
Thinking through some scenarios:

1. storage encounters returns an error to adapter after applying some commands. adapter retries. commands are idempotent so everything's fine. end state is all tables share the same closed timestamp.
2. storage crashes after applying some commands. adapter, uh, times out (? does it? i really don't know), then retries again. still fine because adapter retries and it's idempotent. end state is all tables share the same closed timestamp.
3. storage sends command, adapter applies some. both adapter and storage crash and restart. adapter doesn't have the writes in memory anymore. this differs from the above because not all tables will share the same closed timestamp. does this matter? i'm not sure it does. after restart, the one second close loop we have will eventually bump all tables to have the same timestamp closed. the timestamp oracle still provides linearizability. the closing of all tables when there's a write to one table isn't needed for correctness, but to make queries over any table available at a timestamp.

I think that means this PR is fine as is, and we can change things in the future to maybe...have the storage append message be only for a single table, we disable multi table txns in adapter, and we add a CloseAllTablesAtTimestamp message?

Reply:
An example problem with the current code as is: 3 could also result in a partially applied multi table transaction on ill timed restart, so I think we absolutely have to disallow those.

Reply:
Re-reading this PR I'm now way less sure of my LGTM. I wonder if we should instead rename Insert to Append, have it still take only a single table, but it also does a capability downgrade? (And keep the advance all inputs message.)